Add ViewAreProhibited and ExternalKeyVault FLE tests.

DmitryLukyanov opened the pull request:
Evergreen: https://evergreen.mongodb.com/version/5d9666762fbabe1c34b04830

Prose tests:
https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/README.rst#external-key-vault-test
https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/README.rst#views-are-prohibited
Kevin's repo: https://github.com/kevinAlbs/JavaFLEProsePoC/tree/master/src/main/java

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -0,0 +1,106 @@
+                        && !System.getProperty("org.mongodb.test.awsAccessKeyId").isEmpty());
+
+        MongoClientSettings.Builder clientSettingsBuilder = getMongoClientSettingsBuilder();
+        CodecRegistry codecRegistry = fromRegistries(
```
Where did you get this from?  I don't think it's necessary to alter the UuidCodec if we're always using BsonDocument.  Try removing it and see if the tests still pass.  If so, I'm not sure if you need a new MongoClient for this test at all.  You can just grab the global one from the Fixture.

DmitryLukyanov replied:
from Kevin's repo: https://github.com/kevinAlbs/JavaFLEProsePoC/blob/master/src/main/java/ViewTest.java#L24

Done.

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -0,0 +1,106 @@
+                .append("create", new BsonString("view"))
+                .append("viewOn", new BsonString("coll"));
+
+        db.runCommand(createCmd, BsonDocument.class);
```
There is a createView helper on MongoDatabase

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -0,0 +1,106 @@
+        AutoEncryptionSettings autoEncryptionSettings = autoEncryptionSettingsBuilder.build();
+
+        clientSettings = clientSettingsBuilder
+                .codecRegistry(codecRegistry)
```
This is redundant at best, and probably unnecessary both here and above.

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -0,0 +1,106 @@
+public class ClientSideEncryptionViewAreProhibitedTest {
+    private MongoClient client, clientEncrypted;
+
+    @Before
```
This needs a @After annotated method to close the MongoClient instance(s)

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -0,0 +1,106 @@
+            exceptionThrown = true;
+            assertTrue(me.getMessage().contains("cannot auto encrypt a view"));
+        }
+        assertTrue(exceptionThrown);
```
The more traditional way of doing this is to put a call to `fail()` inside the try block, right after the call to `insertOne`.  This will assert when called, which is what you want. Then you can remove the boolean and this assertion.

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -0,0 +1,170 @@
+                System.getProperty("org.mongodb.test.awsAccessKeyId") != null
+                    && !System.getProperty("org.mongodb.test.awsAccessKeyId").isEmpty());
+
+        CodecRegistry codecRegistry = fromRegistries(
```
I don't think we need a customer CodecRegistry and therefore, don't need this MongoClient at all.  Can just use the default.

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -0,0 +1,170 @@
+        MongoClientSettings externalClientSettings = null;
+        if (withExternalKeyVault) {
+            externalClientSettings = clientSettingsBuilder
+                    .codecRegistry(codecRegistry)
```
Again, unnecessary

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -0,0 +1,170 @@
+        this.withExternalKeyVault = withExternalKeyVault;
+    }
+
+    @Before
```
Need an @After annotated method to close all these Closeable objects

DmitryLukyanov replied:
done

DmitryLukyanov reviewed (commented):
Evergreen: https://evergreen.mongodb.com/version/5d97d2f83e8e8663acdb499b

DmitryLukyanov commented on driver-async/src/test/functional/com/mongodb/async/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -0,0 +1,187 @@
+    @After
+    public void after() {
+        if (clientEncrypted != null) {
+            try {
```
probably we need to create a ticket to investigate why we have `failed` closing sometimes.

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -78,15 +70,10 @@ public void setUp() throws IOException, URISyntaxException {
 
         /* Step 1: create unencrypted client and recreate keys collection */
-        this.client = MongoClients.create(clientSettings);
+        client = Fixture.getMongoClient();
```
Nit: statically import this method

DmitryLukyanov replied:
done

DmitryLukyanov replied:
done

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -78,15 +70,10 @@ public void setUp() throws IOException, URISyntaxException {
-                .build();
+        MongoClientSettings.Builder clientSettingsBuilder = getMongoClientBuilderFromConnectionString();
 
         /* Step 1: create unencrypted client and recreate keys collection */
```
Nit: we're no longer creating it

DmitryLukyanov replied:
done

DmitryLukyanov replied:
done

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -117,16 +103,14 @@ public void setUp() throws IOException, URISyntaxException {
 
-        clientSettings = clientSettingsBuilder
-                .codecRegistry(codecRegistry)
+        MongoClientSettings clientSettings = clientSettingsBuilder
```
The builder is mutable.  So by modifying it here (adding autoEncryptionSettings) that modification remains when it's reused below.  I suggest just getting a new one for each usage.

DmitryLukyanov replied:
done

DmitryLukyanov replied:
done

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -57,25 +52,16 @@ public void setUp() {
-        clientSettingsBuilder.codecRegistry(codecRegistry);
-        MongoClientSettings clientSettings = clientSettingsBuilder.build();
-        client = MongoClients.create(clientSettings);
+        MongoClient client = Fixture.getMongoClient();
```
Nit: statically import

DmitryLukyanov replied:
done

DmitryLukyanov replied:
done

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -57,25 +52,16 @@ public void setUp() {
                         && !System.getProperty("org.mongodb.test.awsAccessKeyId").isEmpty());
 
         MongoClientSettings.Builder clientSettingsBuilder = getMongoClientBuilderFromConnectionString();
```
Move closer to first use

DmitryLukyanov replied:
done

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -73,16 +69,10 @@ public void setUp() throws IOException, URISyntaxException {
-        CodecRegistry codecRegistry = fromRegistries(
-                fromCodecs(new UuidCodec(UuidRepresentation.STANDARD)),
-                MongoClientSettings.getDefaultCodecRegistry());
         MongoClientSettings.Builder clientSettingsBuilder = getMongoClientSettingsBuilder();
```
Move closer to first use

DmitryLukyanov replied:
I replaced it on `getMongoClientSettingsBuilder` as suggested in the `async` test

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionExternalKeyVaultTest.java:
```diff
@@ -73,16 +69,10 @@ public void setUp() throws IOException, URISyntaxException {
 
         /* Step 1: create unencrypted client and recreate keys collection */
-        this.client = MongoClients.create(clientSettings);
+        client = Fixture.getMongoClient();
```
Statically import

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -52,21 +51,12 @@ public void setUp() {
-        clientSettingsBuilder.codecRegistry(codecRegistry);
-        MongoClientSettings clientSettings = clientSettingsBuilder.build();
-        client = MongoClients.create(clientSettings);
+        MongoClient client = Fixture.getMongoClient();
```
Statically import

DmitryLukyanov replied:
done

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionViewAreProhibitedTest.java:
```diff
@@ -52,21 +51,12 @@ public void setUp() {
-                .append("viewOn", new BsonString("coll"));
-
-        db.runCommand(createCmd, BsonDocument.class);
+        db.createView("view", "coll", asList(Document.parse("{$match: {}}")));
```
Can this just be Collections.emptyList()?

DmitryLukyanov replied:
I will check it tomorrow.

DmitryLukyanov replied:
I used this one: `Collections.<BsonDocument>emptyList()` (`Collections.emptyList()` probably is not compliled)

jyemin reviewed (changes requested):
Looking good.  Just a bunch of small suggestions.

DmitryLukyanov reviewed (commented):
https://github.com/DmitryLukyanov/mongo-java-driver/pull/1/files#r332771879
this comment is not implemented yet.

Evergreen:  https://evergreen.mongodb.com/version/5d9d1ea757e85a6a32aef566

DmitryLukyanov commented:
Latest evergreen:  https://evergreen.mongodb.com/version/5d9dc94530661568ece216ba

jyemin reviewed (approved):
Everything looks good now.  I don't see a way to resolve the conversations as I do normally, so just leaving them open.

LGTM

jyemin commented:
Do you plan to merge this soon?

DmitryLukyanov commented:
Do you mean merging into my fork? Probably I don't have rights to push changes in the upstream.

DmitryLukyanov committed:
Add ViewAreProhibited and ExternalKeyVault FLE tests.

DmitryLukyanov commented:
I squashed changes and rebased my branch on the latest upstream master: https://github.com/DmitryLukyanov/mongo-java-driver/commit/6539b570134b9f791d127eab56f51e7c7b742291

jyemin commented:
OK, I will merge them

DmitryLukyanov commented:
Merged to master.
